### PR TITLE
fetch logs endpoint using fetch token String

### PR DESCRIPTION
**Pull request by calvinrp**
This changes the `POST /v1/fetch/logs` endpoint to use a server implementation defined `fetchToken` for the last known record instead of the `RecordId`. It still can be the `RecordId`, but it doesn't need to be. In fact, the reference implementation still uses the `RecordId` for the `fetchToken`.

The motivation here is to give server implementations greater flexibility on implementing the fetch logs endpoint.

Also, in this PR, the Client CLI will be more resilient in handling records that are returned that it has already seen before.

### CODE REVIEW

File: crates/api/src/v1/fetch.rs [hunk 1]
```diff
@@ -53,8 +64,8 @@ pub enum FetchError {
     #[error("log `{0}` was not found")]
     LogNotFound(LogId),
     /// The provided record was not found.
-    #[error("record `{0}` was not found")]
-    RecordNotFound(RecordId),
+    #[error("record fetch token `{0}` was not found")]
+    RecordNotFound(String),
```
Comment (lann) [inline crates/api/src/v1/fetch.rs:68]: Maybe `InvalidFetchToken`? Also do we even need this? Now that the client can handle it we might just want to return the entire log on invalid fetch token to allow changes to the token format.

File: crates/api/src/v1/fetch.rs [hunk 2]
```diff
@@ -6,10 +6,21 @@ use std::{borrow::Cow, collections::HashMap};
...
 use warg_protocol::{
-    registry::{LogId, RecordId, RegistryLen},
+    registry::{LogId, RegistryLen},
     PublishedProtoEnvelopeBody,
 };
 
+/// Wraps the PublishedProtoEnvelopeBody with a fetch token.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PublishedRecord {
```
[the remark this answers is not in the document]
Comment (calvinrp) [reply crates/api/src/v1/fetch.rs:16]: Conceptually, I thought `fetchToken` is part of the `api` crate and not the `protocol` crate (where the `PublishedProtoEnvelopeBody` lives). Also, the `PublishedProtoEnvelopeBody` is used for the `GET /v1/package/{logId}/record/{recordId}` endpoint. The `fetchToken` is only relevant for the `POST /v1/fetch/logs`.

Thought a wrapper made the most sense. But let me know your thoughts on this.

File: crates/api/src/v1/fetch.rs (hunk 1 above)
Comment (calvinrp) [reply crates/api/src/v1/fetch.rs:68]: I think an error makes sense. Happy to rename it to `InvalidFetchToken`.

Comment (lann) [reply crates/api/src/v1/fetch.rs:68]: How would a server migrate fetch token formats?

Comment (calvinrp) [reply crates/api/src/v1/fetch.rs:68]: With this PR, as is:

TLDR: it would work just fine.

The server implementation isn't required to return an error for `InvalidFetchToken`. If it does not error and returns more records than necessary, the client will ignore the records that it already has (without updating to the new `fetchToken` for the previous records). Once a new record is encountered, that new `fetchToken` is now saved and used for future pagination.

I would assume that a server implementation that is migrating to a new `fetchToken` format would not error and would still support the old format (detecting the difference if format versions) for some period of time.

Comment (calvinrp) [reply crates/api/src/v1/fetch.rs:68]: I'll see if I can push a change that makes the client more resilient for `fetchToken` migration.

Comment (calvinrp) [reply crates/api/src/v1/fetch.rs]: We can discuss on the Registry call, but this implementation is likely sufficiently resilient as is.

Comment (calvinrp) [reply crates/api/src/v1/fetch.rs]: Renamed the `RecordNotFound` for the `FetchError` to `FetchTokenNotFound`. Hope that feels better.

File: crates/api/src/v1/fetch.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lann) [reply crates/api/src/v1/fetch.rs:16]: Not I big deal for now. Just a lot of layers of indirection at this point.

[the remark this answers is not in the document]
Comment (calvinrp) [reply crates/api/src/v1/fetch.rs:16]: Yeah, agree. Probably ok for now.

File: crates/client/src/lib.rs [hunk 3]
```diff
@@ -405,12 +399,21 @@ impl<R: RegistryStorage, C: ContentStorage> Client<R, C> {
...
-                    .validate(&record.envelope)
-                    .map_err(|inner| ClientError::OperatorValidationFailed { inner })?;
-                operator.head_registry_index = Some(record.registry_index);
+                let proto_envelope: PublishedProtoEnvelope<operator::OperatorRecord> =
+                    record.envelope.try_into()?;
+                if operator.head_registry_index.is_none()
+                    || operator.head_registry_index.is_some_and(|registry_index| {
+                        proto_envelope.registry_index > registry_index
+                    })
+                {
```
Comment (lann) [inline crates/client/src/lib.rs:408]: Alternative wording:

```suggestion
                if proto_envelope.registry_index > operator.head_registry_index.unwrap_or(0) {
```

File: crates/client/src/lib.rs [hunk 4]
```diff
@@ -405,12 +399,21 @@ impl<R: RegistryStorage, C: ContentStorage> Client<R, C> {
...
+                    // skips over records that has already seen
```
Comment (lann) [inline crates/client/src/lib.rs:409]: Move outside of the `if`?

File: crates/client/src/lib.rs [hunk 5]
```diff
@@ -419,15 +422,24 @@ impl<R: RegistryStorage, C: ContentStorage> Client<R, C> {
...
-                    })?;
-                    package.head_registry_index = Some(record.registry_index);
+                    let proto_envelope: PublishedProtoEnvelope<package::PackageRecord> =
+                        record.envelope.try_into()?;
+                    if package.head_registry_index.is_none()
+                        || package.head_registry_index.is_some_and(|registry_index| {
+                            proto_envelope.registry_index > registry_index
+                        })
+                    {
+                        // skips over records that has already seen
```
Comment (lann) [inline crates/client/src/lib.rs:432]: Ditto above.

File: crates/server/openapi.yaml [hunk 6]
```diff
@@ -802,17 +805,22 @@ components:
...
             - registryIndex
+            - fetchToken
           properties:
             registryIndex:
               type: integer
               description: The index of the published record in the registry log.
               example: 42
+            fetchToken:
+              type: string
+              description: The fetch token for the registry log. Used for pagination in the fetch logs endpoint.
```
Comment (lann) [inline crates/server/openapi.yaml:822]: :shrug: 
```suggestion
              description: The fetch token for the registry log. Used for incremental sync in the fetch logs endpoint.
```

**Review by lann — APPROVED** [review]
A few style/wording nits. Nothing blocking.

File: crates/server/openapi.yaml (hunk 6 above)
Comment (calvinrp) [reply crates/server/openapi.yaml:822]: `Used as a cursor for incrementally fetching log records.` ?

File: crates/client/src/lib.rs (hunk 3 above)
Comment (calvinrp) [reply crates/client/src/lib.rs:408]: Thanks for the suggestion, still getting better at Rust. I'll fix. Doesn't quite work I think given it is an unsigned integer and `0` is a valid (and likely registry index for the Operator log). Seeing if it can be cleaner.

Comment (calvinrp) [reply crates/client/src/lib.rs:408]: So I can simplifiy it, to `operator.head_registry_index.is_none() || proto_envelope.registry_index > operator.head_registry_index.unwrap()`. Since the `.is_none() || ` will ensure the `.unwrap()` won't panic.

**Review by lann — APPROVED** [review]
